module/userExtension/helper.js

```diff
@@ -211,304 +213,400 @@ module.exports = class UserExtensionHelper {
+
+                      for (const role of newRoles) {
+                        if (!currentRoles.includes(role)) {
+                          existingUserProgramRoleMapping[currentRoleInfoIndex].roles.push(role);
+                          kafkaEventPayloads.push({
+                            userId: userProfile.id,
+                            username: userProfile.username,
+                            programId: programId,
+                            role,
+                            eventType: messageConstants.common.CREATE_EVENT_TYPE,
+                          });
+                        }
+                      }
+                    } else {
+                      // Create new program entry
+                      existingUserProgramRoleMapping.push({
+                        programId: programId,
+                        roles: [...newRoles],
+                      });
+
+                      // All roles are new, emit create events
+                      for (const role of newRoles) {
+                        kafkaEventPayloads.push({
+                          userId: userProfile.id,
+                          username: userProfile.username,
+                          programId: programId,
+                          role,
+                          eventType: messageConstants.common.CREATE_EVENT_TYPE,
+                        });
+                      }
+                    }
+                  } else if (userRole.programOperation === messageConstants.common.REMOVE_OPERATION) {
+                    if (currentRoleInfoIndex !== -1) {
+                      const currentRoles = existingUserProgramRoleMapping[currentRoleInfoIndex].roles;
+
+                      // Remove specified roles
+                      const rolesToKeep = currentRoles.filter((role) => !newRoles.includes(role));
+
+                      // Emit delete events for removed roles
+                      const rolesToRemove = currentRoles.filter((role) => newRoles.includes(role));
+                      for (const role of rolesToRemove) {
+                        kafkaEventPayloads.push({
+                          userId: userProfile.id,
+                          username: userProfile.username,
+                          programId: programId,
+                          role,
+                          eventType: messageConstants.common.DELETE_EVENT_TYPE,
+                        });
+                      }
+
+                      if (rolesToKeep.length === 0) {
+                        // Remove entire program entry if no roles left
+                        existingUserProgramRoleMapping.splice(currentRoleInfoIndex, 1);
+                      } else {
+                        // Update with remaining roles
+                        existingUserProgramRoleMapping[currentRoleInfoIndex].roles = rolesToKeep;
+                      }
+                    }
+                    // If program doesn't exist, nothing to remove - no error needed
```

Comment:
@borkarsaish65 if program is not there we should add message in csv right ?
otherwise how they will know they were mapping to a non existing program